picoquic/bytestream.c

```diff
@@ -0,0 +1,306 @@
+/*
+* Author: Christian Huitema
+* Copyright (c) 2018, Private Octopus, Inc.
+* All rights reserved.
+*
+* Permission to use, copy, modify, and distribute this software for any
+* purpose with or without fee is hereby granted, provided that the above
+* copyright notice and this permission notice appear in all copies.
+*
+* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+* ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+* WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+* DISCLAIMED. IN NO EVENT SHALL Private Octopus, Inc. BE LIABLE FOR ANY
+* DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+* (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+* LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+* ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+* (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+* SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#include <stdlib.h>
+#include <string.h>
+
+#include "bytestream.h"
+#include "picoquic_internal.h"
+
+static int bytestream_error(bytestream * s);
+
+bytestream * bytereader_init(bytestream * s, const void * bytes, size_t nb_bytes)
+{
+    s->data = (uint8_t*)bytes;
+    s->size = nb_bytes;
+    s->ptr = 0;
+
+    return s;
+}
+
+bytestream * bytewriter_init(bytestream_buf * s)
+{
+    s->s.data = s->buf;
+    s->s.size = BYTESTREAM_MAX_BUFFER_SIZE;
+    s->s.ptr = 0;
+
+    return &s->s;
+}
+
+bytestream * bytestream_alloc(bytestream * s, size_t nb_bytes)
+{
+    s->data = (uint8_t*)malloc(nb_bytes);
+    if (s->data == NULL) {
+        free(s);
+        return NULL;
+    }
+    s->size = nb_bytes;
+    s->ptr = 0;
+
+    return s;
+}
+
+void bytestream_delete(bytestream * s)
+{
+    if (s->data != NULL) {
+        free(s->data);
+        s->data = NULL;
+    }
+}
+
```

Comment:
If we are going to have separate byte streams for each packet in the file, you may consider some kind of "recycle" structure, similar to what we do for packet buffers. That will cut down on the number of "malloc".

Reply:
I introduced the version with the 2048 byte buffer for that, so that no malloc is necessary at all. The malloc initialized bytestreams are currently only used in the conversion tool for reading/decoding the byte streams again.